### DISCUSSION TITLE
AutoKey on Wayland - Features And Bugs

### DISCUSSION THREAD

**Discussion by josephj11** [autokey/autokey#866] [Maintenance goals]
This is a catch all discussion for the new Wayland interface code.

**Comment by josephj11** [comment]
My initial questions:

What works and what doesn't (as in isn't implemented (yet))?

How does this get invoked? Does it just kick in when it detects Wayland is the display server?
Either way - at least for now - this should be logged so we have some idea of what we're looking at when someone submits a trace.

Linux has a daemon or something watchdog, oomreaper... that watch for lockups and take some action. Is there anything analogous to that we can do in AutoKey for potential lockups?

See PR #844 for the code being discussed here. As of 05/04/2024 Sam has been silent for quite a few months, but AFAIK, his code is still available.

**Reply by sebastiansam55** [reply]
> What works and what doesn't (as in isn't implemented (yet))?

Well, the better question is what _does_ AutoKey do? If you have a complete list of features I can go down the list and tell you the status of them. 

Most user API's should work without issue, `keyboard.wait_for_` calls are not implemented for sure.

`mouse` has not been implemented with wayland either.

The `window` API is all (99%) written on the back of `wmctrl` which is X11 specific last time I checked so I would not expect any API calls in that to work as expected with wayland currently. 

> How does this get invoked? Does it just kick in when it detects Wayland is the display server?

Yes, it checks the `$XDG_SESSION_TYPE`, if that is set to `wayland`, it will try to load the uinput_interface instead of the x11 one.


RE: Autokey hanging/system lockups
Obviously a major concern. Personally I'd recommend any testing should be done in VMs to avoid issues like that in the "pre-alpha" state currently.

Once the process dies it should release the keyboard grab, I'll be putting more thought into how to handle failures at this level, I think you'd also run into issues if the keyboard was unplugged while AK is running right now which we should attempt to handle gracefully.

**Comment by josephj11** [comment]
@sebastiansam55 What do you notice that's different about Wayland in general from a user perspective?

I have only looked into it a little bit and most of what I saw was about enhancing security and getting rid of ancient code - important, but not really user facing.

**Reply by sebastiansam55** [reply]
Ideally there should not be a difference to the AutoKey end user. 

In practice there are many in the current version of that PR;
* Simplistic device grabbing (searches the device description for `keyboard` and `mouse` respectively
* Special characters (like F13 which is written to the config file when used as a hotkey as `<code193>`, this code is like an x11 keysym code, not sure that there is a way to translate this to something uinput can interpret without x11 libraries.
* I've not tested most API calls aside from the `keyboard` module at this point but `dialog`, `clipboard` and `highlevel` should all work the same across x11/wayland

> most of what I saw was about enhancing security 
Speaking of security, I'm not sure what specific thing you're referring to here but there are security considerations that users will have to make with the new version of autokey as it currently exists, it relies on the user being in the `input` group which is not the case by default. 

I'm not 100% of the total implications of this but in effect it allows a non-root process to send input via the mouse/keyboard and other input devices (which is what we want so it's a tradeoff).

Additionally, I don't think there is really a way to secure DBus methods so with the gnome-shell window extension as it is any process would be able to poll the windows and interact with them using those DBus methods. 

Whether those are serious security concerns is another question altogether.

**Comment by josephj11** [comment]
@sebastiansam55 Did you ever consider writing a priviledged AutoKey daemon that would handle the just the interface I/O? I wonder if that would make anything easier.

**Comment by Elliria** [comment]
Hopefully some of our Wayland users will test this code out.

**Comment by josephj11** [comment]
Where does this whole enhancement fit? Would it need to go into 0.97.0?

**Comment by josephj11** [comment]
@sebastiansam55 Questions:

How does this impact multibyte character sets (anything other than EN US)? Will the new keyboard.send_keys() just work with them?

If this interface will work under X11, we may want to add an option to enable it when X11 is detected.

We just had a strange report on our Google group about a mouse problem. (I asked the OP to file an issue.) If that can't be figured out/fixed quickly, then the user could try out this new interface to see if it helps.

Is there any value in offering a mix and match strategy so an X11 user could keep our normal API, but also explicitly use the uinput API calls as needed?

**Comment by LinuxBoy-96** [comment]
I'm using Cosmic, so Wayland only.
Let me tell you this, Autokey is unusable right now. 
It cannot detect the key when you set a new Hotkey.
I can delete the default folders but if I create a new one; Autokey crash and everything come back when I  open it again.
I guess it's normal since Autokey is for X11. but that's my point how can it still not work on wayland in 2024?
I'm flabberguasted.
I use Autokey Qt

**Comment by Elliria** [comment]
If you go to the page for [pull request #844](https://github.com/autokey/autokey/pull/844), that's where work has been done on Wayland compatibility. The pull request is still in Draft mode, but the comments suggest that it's at least mostly working under Wayland. Testers for it are needed and welcome.

[This comment on that page](https://github.com/autokey/autokey/pull/844#issuecomment-1538846177) gives instructions for cloning that pull request and running it on your system without installing it and without interfering with your existing AutoKey installation (you'll be running the cloned copy of AutoKey from wherever you put it).

**Comment by josephj11** [comment]
This project runs on volunteer efforts.Volunteers work on whatever interests them. We can only control what gets merged.We have never had more than one main developer and never more than two active contributors. Check out issue #87. No one implemented it in all this time. There is a partial implementation (keyboard and mouse modules) referenced in this issue or in a link.We have no currently active developers to do anything. If you can help to solve that it would  be immensely appreciated.We're listed at codetriage.com and are looking into Opire for bounties.OT: you're the first person I've encountered who's using Cosmic. Love to hear about that over on Gitter.im/autokey/autokey. (I'm on my phone, so the link might not be right.)Sent from my Galaxy
-------- Original message --------From: LinuxBoy-96 ***@***.***> Date: 9/5/24  3:29 PM  (GMT-05:00) To: autokey/autokey ***@***.***> Cc: Joe ***@***.***>, Author ***@***.***> Subject: Re: [autokey/autokey] AutoKey on Wayland - Features And Bugs (Discussion #866) 
I'm using Cosmic, so Wayland only.
Let me tell you this, Autokey is unusable right now.
It cannot detect the key when you set a new Hotkey.
I can delete the default folders but if I create a new one; Autokey crash and everything come back when I  open it again.
I guess it's normal since Autokey is for X11. but that's my point how can it still not work on wayland in 2024?
I'm flabberguasted.
I use Autokey Qt

**Comment by LinuxBoy-96** [comment]
Ok so I have A hot take.
I would like to take a look what could be done to have a good AutoKey on wayland but there is a few thing to say about that

1. For simplicity sake, the wayland Autokey will be diffrent source than X11 Autokey.
2. That version will use the Iced GUI.
3. That version being written in 100% Rust, (explain #2)

Why Rust?:
- AutoKey possibly deal will sensible macros and passwords.
- It may save some resources.
- I think, at the end, this is a better path, same with the idea to think about a wayland only version, X11 won't stay for ever.
- It's the only thing I think I could do, sorry :(

**Comment by josephj11** [comment]
Welcome to the AutoKey community, Linuxboy-96!We'd love to have this project moving again!TL;DR: Before such a major effort, demo programs showing that each required feature works at all in Wayland need to be implemented. Wayland's security model may prevent AutoKey from doing things it needs to do.Several things: We have the keyboard and mouse modules working in Wayland, so that's not an issue.You (or anyone working on this) have to demonstrate that hotkeys and window commands, etc. will work in any language even standalone before doing a lot of coding work.All our user scripts are in Python and have to be supported regardless of what language the application itself is written in.I have never looked at rust, but I have only heard good things about it, but if someone does a rewrite they would be introducing a whole new set of potential bugs and would have to commit to long term support to make it safe for us to adopt it. This would be true even if the rewrite was in Python.If you're starting from scratch, I see no reason why the new program should support x11. We have the current one for that.Also, although there may be one or two former developers lurking here, you might not be able to get answers to your questions about the current implementation and you would  have to do all the git, GitHub, and any CI/CD configurations too. Currently, we do not have any active developers to assist you.There is one junior developer who can do a lot to get you started and one former lead developer who might answer some questions.Of course, the project is GPL, so you can fork it and do almost anything you like without our permission. The only issue would be our cooperation with your efforts.Sent from my Galaxy
-------- Original message --------From: LinuxBoy-96 ***@***.***> Date: 9/7/24  7:20 PM  (GMT-05:00) To: autokey/autokey ***@***.***> Cc: Joe ***@***.***>, Author ***@***.***> Subject: Re: [autokey/autokey] AutoKey on Wayland - Features And Bugs (Discussion #866) 
Ok so I have A hot take.
I would like to take a look what could be done to have a good AutoKey on wayland but there is a few thing to say about that

For simplicity sake, the wayland Autokey will be diffrent source than X11 Autokey.
That version will use the Iced GUI.
That version being written in 100% Rust, (explain #2)

Why Rust?:

AutoKey possibly deal will sensible macros and passwords.
It may save some resources.
I think, at the end, this is a better path, same with the idea to think about a wayland only version, X11 won't stay for ever.
It's the only thing I think I could do, sorry :(
